### ISSUE TITLE
sparklyr invocation chooses incorrect overloaded method

### ISSUE THREAD

**Issue by joveyuan-db** [sparklyr/sparklyr#3360]
sparklyr's invocation does not correctly handle overloaded methods with `Array` and `Map` parameter types.  Specifically, in the following example
```
class C {
  def f(args: Array[_]): Unit = {
    println(args.nonEmpty)
  }

  def f(args: Map[String, Any]): Unit = {}
}

val index = findMatchedSignature(classOf[C].getMethods.filter(_.getName == "f").map(_.getParameterTypes), Array(Map()))
classOf[C].getMethods.filter(_.getName == "f")(index.get).invoke(new C, Map())
```
`findMatchedSignature` chooses the first method and we invoke `f(args: Array[_])` with `Map()`.  Checking whether the array is non-empty then throws a `MapError`.  This issue is not exclusive to the fact that the map is empty.

<img width="1087" alt="image" src="https://github.com/sparklyr/sparklyr/assets/67606568/52cdf74e-3f6e-4edf-93bd-3eada1e739b4">

---

I believe this is the root cause of why parameterized queries break on Databricks (https://github.com/sparklyr/sparklyr/blob/4d49e40c8cefd394358f64261fceeaeaef13f8cc/R/dbi_spark_result.R#L72-L73), because Spark 3.5 further overloads the `sql` method (https://github.com/apache/spark/blob/eb4152dfcd6e9ef42750e7bac5238efe641458d2/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L641-L660).  There now exists both a `sql(sqlText: String, args: Array[_])` and a `sql(sqlText: String, args: Map[String, Any])` method signature in newer Spark versions, leading to sparklyr choosing the wrong one.

**Comment by edgararuiz** [comment]
Hi, is this using `sparklyr` dev? or CRAN?

**Comment by joveyuan-db** [comment]
This was using sparklyr 1.8.2, although I suspect it affects other sparklyr versions including dev.

I copied the `findMatchedSignature` method from https://github.com/sparklyr/sparklyr/blob/4d49e40c8cefd394358f64261fceeaeaef13f8cc/java/spark-1.5.2/invoke.scala#L17-L133 into a Scala REPL and was able to confirm the issue on the toy example above.

**Comment by edgararuiz** [comment]
Ok thanks, is there some R code we can run to test that? I don't see an error that we can recreate

**Comment by joveyuan-db** [comment]
Yes, apologies.  I reproduced this in Databricks with Spark 3.5 and sparklyr 1.8.2 with the following query.  The third line is an implementation detail of `sparklyr` (eventually called by `sdf_len`, for example), but isolates the core issue.
```
library(sparklyr)
sc <- spark_connect(method = "databricks")
DBI::dbSendQuery(sc, "SELECT 0L as `id`")
```
<img width="1423" alt="image" src="https://github.com/sparklyr/sparklyr/assets/67606568/794e4864-326d-421a-bb02-44a3660279df">

I haven't gotten the chance to verify this outside of Databricks, but I believe any environment with Spark 3.5 will be sufficient.

**Comment by joveyuan-db** [comment]
Hi Edgar, yeah, I'm testing a preview build internally in Databricks.  I'm using sparklyr from the Databricks cluster itself, so it doesn't require Databricks Connect though.

The root cause (bug in `findMatchedSignature`) is not specific to Spark 3.5 by the way, but 3.5 surfaces it.
